### DISCUSSION TITLE
Триггеры от Tasker (android)

### DISCUSSION THREAD

**Discussion by Chimildic** [Chimildic/goofy#124] [Избранное]
## Описание
Есть несколько вариантов того, как запустить функцию goofy: [в редакторе кода или триггером по времени](https://chimildic.github.io/goofy/#/first-playlist), а также [через веб-страницу](https://github.com/Chimildic/goofy/discussions/9). 

Теперь доступен Tasker - это **андроид** приложение, позволяющее запускать действия по событиям телефона. В нашем случае, действием будет запуск функции goofy, а события выступят триггерами. К примеру, ваше местоположение (дом, работа), нажатие кнопки, открытие приложения Spotify, любое уведомление, будильник, даже голос - список очень большой.

Допустим у вас стандартный триггер по времени на [шаблон любимо и забыто](https://chimildic.github.io/goofy/#/template?id=Любимо-и-забыто). Как установить триггер по местоположению? Код не меняем. Сначала настроим Apps Script и Tasker (один раз), а затем установим сам триггер.

> Насколько я понимаю, Tasker будет срабатывать с некоторой периодичностью, пока вы находитесь в заданном месте (в случае триггера по месту). Чтобы это происходило только раз, нужно предусмотреть условие (в задаче или коде goofy). Пока для этого нет примера. Воздержитесь от настройки, если не знакомы с Tasker настолько, чтобы сделать это самостоятельно.

## Подготовка
- Чтобы не уйти со страницы инструкции, открывайте ссылки в новой вкладке. Через меню правой кнопкой мыши или нажатием на колесико. 
- Необходимо установить Tasker, предпочтительно свежую версию. Скачайте бесплатно [мод на 4PDA](https://4pda.to/forum/index.php?s=&showtopic=173935&view=findpost&p=108086774) или купите в [Google Play](https://play.google.com/store/apps/details?id=net.dinglisch.android.taskerm). Также доступна по подписке Pass там же (1 месяц бесплатно).
- Настройка требует передачи текста с компьютера в телефон, подготовьтесь. Например, [генератор QR-кода](https://kazuhikoarase.github.io/qrcode-generator/js/demo/), сообщение самому себе ВКонтакте, закладки в телеграме или любой другой способ. Писать вручную плохая затея.

## Настройка Apps Script
1. Зайдите в [Google Cloud Platform](https://console.cloud.google.com/). Если делаете это первый раз, появится всплывающее окно. Выберите страну, поставьте первую галку (_I argree_) и нажмите кнопку _agree and continue_. Сверху будет баннер про бесплатный период (_your free trial is waiting_), нажмите отказаться **dismiss**.
2. Создайте новый проект. Для этого перейдите по [ссылке](https://console.cloud.google.com/projectcreate) или нажмите кнопку _create project_ на странице _home_. Задайте произвольное имя, например _goofy for tasker_.
3. По очереди зайдите [сюда](https://console.cloud.google.com/apis/library/drive.googleapis.com), [сюда](https://console.cloud.google.com/apis/library/sheets.googleapis.com) и [сюда](https://console.cloud.google.com/apis/library/script.googleapis.com). Нажмите кнопку _enable_, дождитесь прогрузки.
4. Перейдите [сюда](https://console.cloud.google.com/apis/credentials) или выберите в левом меню _APIs & Services > Credentials_. Нажмите кнопку _configure consent screen_. 

   - выбираем _external_
   - заполняем поля под звездочками: произвольное имя (goofy for tasker) и ваша gmail-почта дважды
   - нажимаем кнопку _add or remove scopes_, поочередно вставляем следующие ссылки в поле поиска и отмечаем их галками в таблице
      - https://www.googleapis.com/auth/script.scriptapp
      - https://www.googleapis.com/auth/script.external_request
      - https://www.googleapis.com/auth/drive
      - https://www.googleapis.com/auth/spreadsheets
   - нажимаем кнопку _update_, должна появится таблица с этими значениями
   
   ![image width="5%"](https://user-images.githubusercontent.com/40138097/132653818-28568de5-4405-4e94-8cc6-079306f1592f.png)

    - нажимаем _save and continue_, теперь _add users_ и вводим вашу gmail-почту, которая используется для проекта в Apps Script.

5. Зайдите [сюда](https://console.cloud.google.com/apis/credentials/consent) или перейдите APIs & Services > OAuth consent screen > goofy for tasker. Нажмите кнопку `public app`, затем `confirm`. 

> Будет предупреждение `Needs verification`, игнорируем.

6. Возвращаемся в раздел [Credentials](https://console.cloud.google.com/apis/credentials). Нажимаем _create credentials > OAuth client ID_
   - тип _web application_, имя произвольно
   - под заголовком _authorized redirect URIs_ нажимаем _add URI_ и добавляем ссылки: `https://tasker.joaoapps.com/auth.html`, `http://localhost:8080/`
 
7. Переходим на страницу _home > dashboard_ или по [ссылке](https://console.cloud.google.com/home/dashboard). В новой вкладке зайдите в настройки вашего проекта goofy в Apps Script. Скопируйте _project number_ и сохраните.

![image](https://user-images.githubusercontent.com/40138097/132650962-b330d2ad-26e6-422d-a9bb-db229bedb919.png)

8. Определите функцию, которую хотите запустит в первый раз для теста. Выберите существующую или напишите новую. 

9. В Apps Script нажимаем _начать развертывание > управления развертываниями_. Слева по очереди выбираем активные и отправляем в архив. 

   ![image](https://user-images.githubusercontent.com/40138097/132958633-30af7aea-c95c-4029-b45b-9645c671312c.png)

10. Нажимаем _начать развертывание > новое развертывание_. В настройках выбираем веб и API. Добавьте описание _веб и api_. Нажимаем начать.

   ![image](https://user-images.githubusercontent.com/40138097/132958677-02fa75c1-f0c1-4c2d-b574-cd0e6c3d5a47.png)

> Шаги 9 и 10 изменят _id_ активного развертывания. Если вы используете [управление с телефона](https://github.com/Chimildic/goofy/discussions/9), обновите ссылку на новую. Тоже самое касается [хостинга github](https://github.com/Chimildic/goofy/discussions/9#discussioncomment-517316). Зайдите в репозиторий и обновить ее в файле _index.html_. Если используется хостинг, возьмите ссылку от веб-приложения, которая заканчивается на _/exec_. Если без, смените на _/dev_. Ссылка находится в управлении развертываниями.

![image](https://user-images.githubusercontent.com/40138097/132958824-42f57b7e-48aa-4e84-8964-68fbefe5131b.png)

### Важно
Если вы внесли в код правки и хотите запускать его через Tasker, нужно обновить развертывание.

![image](https://user-images.githubusercontent.com/40138097/132670545-568dede4-1bda-4e84-b39d-a96902e12299.png)

## Передача на телефон

Пример на основе [генератора QR-кода](https://kazuhikoarase.github.io/qrcode-generator/js/demo/)

1. Скопируйте в поле на сайте генератора следующий текст

```
Исполняемый файл API


Client ID


Client Secret


Endpoint to get code
https://accounts.google.com/o/oauth2/v2/auth

Endpoint to get refresh token
https://www.googleapis.com/oauth2/v4/token

Scopes
https://www.googleapis.com/auth/script.scriptapp
https://www.googleapis.com/auth/script.external_request
https://www.googleapis.com/auth/drive
https://www.googleapis.com/auth/spreadsheets

Body
{"function":"name"}
```

2. Необходимо добавить значения
   - Исполняемый файл API - ссылка, которую нужно скопировать в Apps Script: _начать развертывание > управление развертываниями_. Выберите в списке _веб и api_. Возьмите ссылку под заголовком _исполняемый файл API_ (заканчивается на `:run`)
   - _Body_ - вместо _name_ в кавычки вписать имя функции из goofy, которую нужно запустить. 
   - Client ID и Secret - две строки. Зайдите в [credentials](https://console.cloud.google.com/apis/credentials). Выберите в таблице строчку без восклицательного знака. На открывшейся странице будут нужные значения.
   
![image](https://user-images.githubusercontent.com/40138097/132661070-dadb4428-f335-4d50-8589-71e057e767c9.png) 

3. После вставки значений, сгенерируйте QR Code, нажав кнопку _update_.
4. Отсканируйте QR Code камерой телефона или [сторонним приложением](https://play.google.com/store/apps/details?id=com.gamma.scan)

## Настройка Tasker

1. Зайдите в настройки Tasker и переключите язык на английский: меню три точки > настройки > интерфейс > внизу выбрать _english_.
2. Вернитесь на главный экран приложения. В меню три точки включите Tasker (нажать enable, чтобы появилось disable).

   ![image](https://user-images.githubusercontent.com/40138097/132663489-acc15e0d-8dea-402a-9663-624f1a4f961f.png)

3. Перейдите во вкладку _tasks_, нажмите кнопку добавить (плюс внизу). Наберите имя _Apps Script OAuth2_, нажмите галку. Открылся экран _task edit_. 

   - Нажмите кнопку добавить. Внизу поле фильтра, наберите _auth_. Из списка выберите _HTTP Auth_. Скопируйте в поля значения из текста, который передали на телефон (_client id, client secret, endpoint get code, endpoint get refresh code, scopes_). В поле _scopes_ будет три ссылки, разделенных пробелом или новой строкой. После этого, нажмите стрелку назад (слева сверху).
   - Добавьте действие _return_, также найдя его в списке по фильтру. В поле _value_ напишите _%http_auth_headers_ или выберите его из списка, который открывается при нажатии иконки напротив названия поля. Галка _stop_ должна стоять.

4. В результате получаем задачу из двух действий. Нажмите внизу кнопку плей (первая слева). Откроется браузер. Войдите или выберите аккаунт гугл, на котором сохранен проект Apps Script. Далее нажмите продолжить и разрешить. Когда вернетесь в Tasker, снова нажмите плей. Браузер не должен открыться. Если это произошло, пройдите авторизацию еще раз.

   > Если вы получили ошибку _403 access denied_, то вернитесь [сюда](https://console.cloud.google.com/apis/credentials/consent) и проверьте, чтобы среди _test users_ была почта, которую вы только что выбрали. И повторите шаг через пару минут.

   ![image](https://user-images.githubusercontent.com/40138097/132665787-c16b1d02-b213-4ef9-83c7-c5bd05a7ed97.png)

## Запуск функции goofy

1. Перейдите во вкладку _tasks_. Добавьте задачу с именем _goofy.test_
2. Добавьте действия, найдя их по фильтру
   - _perform task_. В поле _name_ добавьте _Apps Script OAuth2_, выбрав его через иконку лупы. В поле _return value variable_ напишите _%headers_
   - _http request_. В поле _method_ - выбрать _post_, в _url_ - скопировать из текста, переданного на телефон (ссылка от _исполняемый файл API_), в _headers_ вписать _%headers_, в _body_ скопировать из текста `{"function":"name"}`.
3. Нажмите плей. Если вы все сделали верно, Tasker не покажет ошибок, а функция запустится. Проверяйте результат в логах Apps Script.

> При ошибке _500_ удостоверьтесь, что [это](https://console.cloud.google.com/apis/library/drive.googleapis.com), [это](https://console.cloud.google.com/apis/library/sheets.googleapis.com) и [это](https://console.cloud.google.com/apis/library/script.googleapis.com) включено.

4. Чтобы запустить другую функцию, склонируйте задачу _goofy.test_ (долгий тап по задаче, меню три точки, _clone_) и поменяйте имя функции в поле _body_ из действия _http request_.

## Триггер на функцию

Во вкладке _profiles_ добавьте интересующее событие. Например, _location_. Затем в меню три точки нажмите address, впишите адрес (писать до тех пор, пока не появится в списке). Далее внизу задайте радиус. Нажмите кнопку назад, чтобы перейти в _profiles_. Предложит ввести имя для события, потом выбрать задачу.

**Comment by youhim** [comment]
Приветствую. Сделал настройку по инструкции. Практически все получилось кроме последнего и основного. Таскер не запоминает авторизацию. После запуска задачи я прохожу авторизацию в гугле и на долю секунды появляется окно "Authentication Done! You can close this window". Но после повторного запуска задачи опять просит авторизироваться. И так бесконечно. Потом я настроил горячие клавиши на ПК и там авторизация прошла нормально, все работает. Значит на стороне сервера все настроил правильно. Не понятно почему в Таскере не работает. К тому же вчера я настраивал OAuth2 Таскера у Spotify и оно работает. Значит проблема не с Таскером? Есть идеи?

**Reply by Tavi1977** [reply]
Ваш гугл аккаунт, используемый по-умолчанию, не соответствует аккаунту, на котором сохранен проект Apps Script. Нужно сначала выйти из всех! учетных записей гугл, а затем в первую очередь войти в тот аккаунт, к которому привязан проект. Аккаунт установится в качестве основного, и запрос на повторную авторизацию прекратится
